teps/0044-decouple-task-composition-from-scheduling.md

```diff
@@ -0,0 +1,954 @@
+   runAfter: get-source
+   taskRef:
+     name: just-unit-tests
+   workspaces:
+   - name: source-code
+     workspcae: source-code
+   - name: test-results
+     workspace: test-results
+ - name: upload-results
+   runAfter: run-unit-tests
+   taskRef:
+     name: gcs-upload
+   params:
+   - name: location
+     value: gs://my-test-results-bucket/testrun-$(taskRun.name)
+   workspaces:
+   - name: data
+     workspace: test-results
+finally:
+- name: update-slack
+  params:
+  - name: message
+    value: "Tests completed with $(tasks.run-unit-tests.status) status"
+```
+
+Possible tweak: we could do this scheduling only when
+[a Task requires a workspace `from` another Task](https://github.com/tektoncd/pipeline/issues/3109).
+
+Pros:
+* Doesn't require any changes for Pipeline or Task authors
+
+Cons:
+* Will need to update our entrypoint logic to allow for steps running in parallel
+* Doesn't give as much flexibility as being explicit
+  * This functionality might not even be desirable for folks who want to make use of multiple nodes
+
+### Introduce scheduling rules to pipeline
+
+In these options, we add some syntax that allows Pipeline authors to express how they want Tasks to be executed.
+
+#### Add "grouping" to tasks in a pipeline
+
+In this option we add some notion of "groups" into a Pipeline; any Tasks in a group will be scheduled together.
+
+In this example, everything in the `fetch-test-upload` group would be executed as one pod. The `update-slack` Task would
+be a separate pod.
+
+```yaml
+kind: Pipeline
+metadata:
+  name: build-test-deploy
+spec:
+ params:
+  - name: url
+    value: https://github.com/tektoncd/pipeline.git
+  - name: revision
+    value: v0.11.3
+ workspaces:
+  - name: source-code
+  - name: test-results
+ tasks:
+ - name: get-source
+   group: fetch-test-upload # our new group syntax
+   workspaces:
+   - name: source-code
+     workspace: source-code
+   taskRef:
+     name: git-clone
+   params:
+   - name: url
+      value: $(params.url)
+    - name: revision
+      value: $(params.revision)
+ - name: run-unit-tests
+   group: fetch-test-upload # our new group syntax
+   runAfter: get-source
+   taskRef:
+     name: just-unit-tests
+   workspaces:
+   - name: source-code
+     workspcae: source-code
+   - name: test-results
+     workspace: test-results
+ - name: upload-results
+   group: fetch-test-upload # our new group syntax
+   runAfter: run-unit-tests
+   taskRef:
+     name: gcs-upload
+   params:
+   - name: location
+     value: gs://my-test-results-bucket/testrun-$(taskRun.name)
+   workspaces:
+   - name: data
+     workspace: test-results
+finally:
+- name: update-slack
+  params:
+  - name: message
+    value: "Tests completed with $(tasks.run-unit-tests.status) status"
+```
+
+Pros:
+* Minimal changes for Pipeline authors
+
+Cons:
+* Will need to update our entrypoint logic to allow for steps running in parallel
+  * We could (at least initially) only support sequential groups
```

Comment:
Note that this is a "con" for almost **all** approach that will allow multiple task to run in a pod (in addition to request/limit, result+variable interop changes, …)

Reply:
good point! ive added this under the "alternatives" heading and tried to highlight that this applies to most of the proposals (happy to copy + paste it into multiple alternatives if that's clearer, once we're happy with the wording)